src/migrate/index.js

```diff
@@ -121,22 +131,84 @@ export default class Migrator {
     });
   }
 
+  _createMigrationLockTable(tableName) {
+    return this.knex.schema.createTable(tableName, function(t) {
+      t.boolean('is_locked');
+    });
+  }
+
+  _getLockTableName(tableName) {
+    return tableName + '_lock';
+  }
+
+  _isLocked(trx) {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .transacting(trx)
+      .forUpdate()
+      .select('*')
+      .then(data => {
+        // This is to handle the different databases
+        // data[0].is_locked == false will also work but
+        // JSHint requires === which won't do triple equals
+        if (data[0].is_locked !== false &&
+            data[0].is_locked !== '0' &&
+            data[0].is_locked !== 0) {
+          return true;
+        }
+
+        return false;
+      });
+  }
+
+  _lockMigrations(trx) {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .transacting(trx)
+      .update({ is_locked: true });
+  }
+
+  _unlockMigrations() {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .update({ is_locked: false });
+  }
+
   // Run a batch of current migrations, in sequence.
   _runBatch(migrations, direction) {
-    return Promise.all(_.map(migrations, this._validateMigrationStructure, this))
-      .then(() => this._latestBatchNumber())
-      .then((batchNo) => {
-        if (direction === 'up') batchNo++;
-        return batchNo;
-      })
-      .then((batchNo) => {
-        return this._waterfallBatch(batchNo, migrations, direction)
-      })
-      .catch((error) => {
-        helpers.warn('migrations failed with error: ' + error.message)
-        throw error
-      })
-  }
+    return this.knex.transaction(trx => {
+      return this._isLocked(trx)
+        .then(isLocked => {
+          if (isLocked) {
+            throw new LockError();
+          }
+          return Promise.all(_.map(migrations, this._validateMigrationStructure, this))
```

Comment:
Verifying migration files could be after the locking transaction. It is a bit strange here in a middle of locking.